ledger/eval/eval.go

```diff
@@ -972,17 +978,39 @@ func (eval *BlockEvaluator) TransactionGroup(txgroup []transactions.SignedTxnWit
 
 	// Evaluate each transaction in the group
 	txibs = make([]transactions.SignedTxnInBlock, 0, len(txgroup))
+	cowForTxn := cow
 	for gi, txad := range txgroup {
 		var txib transactions.SignedTxnInBlock
 
+		if eval.GranularEval {
+			if gi == 0 {
+				cowForTxn = cow.child(1)
+				defer cowForTxn.recycle()
+			} else {
+				// Reuse the same cow allocation for all transactions in this group
+				cowForTxn.reset()
+				cow.reuseChild(cowForTxn, 1)
+			}
+		}
+
 		if eval.Tracer != nil {
 			eval.Tracer.BeforeTxn(evalParams, gi)
 		}
 
-		err := eval.transaction(txad.SignedTxn, evalParams, gi, txad.ApplyData, cow, &txib)
+		err := eval.transaction(txad.SignedTxn, evalParams, gi, txad.ApplyData, cowForTxn, &txib)
 
 		if eval.Tracer != nil {
-			eval.Tracer.AfterTxn(evalParams, gi, txib.ApplyData, err)
+			var deltas *ledgercore.StateDelta
+			if eval.GranularEval {
+				// Only include if we're sure the cowForTxn contains ONLY the deltas from this txn
+				d := cowForTxn.deltas()
+				deltas = &d
```

Comment:
seems like here if you make a copy of the deltas, you can use a pool cow and then leave all the recycling code alone?

Reply:
Recycling is already safe, the complexity here is to reduce the number of times we hit the recycling pool, since we know we'll need a cow for each txn. More details in my other comment: https://github.com/algorand/go-algorand/pull/5234/files#r1186414406